### PR TITLE
make: introduce shell-based module dependency system

### PR DESCRIPTION
**Pull request by kaspar030**
Alternative to #3709.

The dependencies can now easily be expressed using simple "used A && use B" logic in shell syntax (it is a shell script). Order doesn't matter as file is evaluated until there's no change anymore.

Single-line dependencies make git's merging algorithm work better.

Downside: it is slow as hell (.5sec on my machine using bash, .3 using dash).

~~(USEPKG's broken, if we want this and accept the slowdown, I'll fix it.)~~

### CODE REVIEW

**Comment by Kijewski** [comment]
I like it. :+1:

**Comment by cgundogan** [comment]
looks pretty and seems to work with zsh

**Comment by kaspar030** [comment]
@authmillenon I don't get the gnrc_ndp_host dependencies. When should it be in?

**Comment by kaspar030** [comment]
- rebased

**Comment by authmillenon** [comment]
@kaspar030 if there is a non-6LoWPAN-interface (similar with gnrc_ndp_router) in #3749

**Comment by kaspar030** [comment]
> @kaspar030 if there is a non-6LoWPAN-interface

So when 

either GNRC_NETIF_NUMOF==1 and sixlowpan is not linked in and gnrc_ndp is selected
or GNRC_NETIF_NUMOF>1 and gnrc_ndp is selected?

**Comment by authmillenon** [comment]
Replace sixlowpan with gnrc_sixlowpan_nd and you got yourself a deal.

**Comment by kaspar030** [comment]
- squashed

**Comment by OlegHahm** [comment]
> Downside: it is slow as hell (.5sec on my machine using bash, .3 using dash).

At first I read "5 sec" and thought "what happened to the Kaspar I know"... ;-)

File: scripts/deps.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
```
Comment (Kijewski) [inline scripts/deps.sh]: set -e

File: Makefile.dep [hunk 2]
```diff
@@ -1,300 +1,12 @@
...
-
-ifneq (,$(filter gnrc_ipv6,$(USEMODULE)))
-  USEMODULE += inet_csum
-  USEMODULE += ipv6_addr
-  USEMODULE += gnrc_ipv6_hdr
-  USEMODULE += gnrc_ipv6_nc
-  USEMODULE += gnrc_ipv6_netif
-endif
+USEMODULE := $(shell GNRC_NETIF_NUMOF=$(GNRC_NETIF_NUMOF) \
+	$(RIOTBASE)/scripts/deps.sh $(RIOTBASE)/modules.dep $(USEMODULE))
```
Comment (Kijewski) [inline Makefile.dep]:
```make
ifeq (,$(USEMODULE))
$(error Something bad happened.)
endif
```

**Comment by kaspar030** [comment]
@Kijewski Hm, you're right, some errorchecking is needed. Unfortunately make doesn't bail out on```$(shell false)```, and the if-statement right after the evaluation is never empty as somewhere we set the default modules.

**Comment by Kijewski** [comment]
@kaspar030, using `set -e` in scripts/deps.sh should kill the script on the first error, so `USEMODULE` should stay empty.

**Comment by kaspar030** [comment]
@Kijewski yup it does. thx!

**Comment by haukepetersen** [comment]
Actually, I see a simple problem: What do windows users do? Is this working for them?

**Comment by cgundogan** [comment]
I guess they will have to use cygwin or any other port of bash and the GNU software tools

**Comment by kaspar030** [comment]
Wow, I didn't think about them windows users at all... Even if cygwin would do it, AFAIR, pipe & spawning subprocesses is slower by factor 10, making this unusable...

**Comment by authmillenon** [comment]
Needs rebase and adaptation to changes in #3749 (and possibly to #3134 later on)

**Comment by authmillenon** [comment]
For windows users: They need make anyway, which includes at least some version of `/bin/sh`. Don't know if it makes sense to introduce an alternative batch/PowerShell script for windows, though.

**Comment by Kijewski** [comment]
There is no subprocesses in your script.
```
$ sh
$ type [
[ is a shell builtin
$ type false
false is a shell builtin
$ type true
true is a shell builtin
$ type .
. is a special shell builtin
```

File: modules.dep [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+used srf02 && use xtimer
+used vtimer && use xtimer timex
+used xbee && use ieee802154
+used xtimer && use FEATURE_periph_timer
+used_filter gnrc_.* gnrc_netapi gnrc_netreg gnrc_netif.* gnrc_pktbuf && use gnrc
+
+used ieee802154 && {
+    used gnrc_ipv6 && use gnrc_sixlowpan
+    used gnrc_ipv6_router && use gnrc_sixlowpan
+    # TODO: replace with gnrc_sixlowpan_router
```
Comment (authmillenon) [inline modules.dep]: done so in #3749

File: modules.dep [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    used gnrc_ipv6_default && use gnrc_sixlowpan_default
+    used gnrc_ipv6_router_default && use gnrc_sixlowpan_default
+    # TODO: replace with gnrc_sixlowpan_router_default
```
Comment (authmillenon) [inline modules.dep]: Dito

**Comment by kaspar030** [comment]
> There are no subprocesses in your script.

tr, sort, grep?

**Comment by kaspar030** [comment]
- rebased

This is hell to rebase...

File: scripts/deps.sh [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+shift
+
+DEPS="$*"
+
+use() {
+    DEPS="$DEPS $*"
+}
+
+used() {
+    echo $DEPS | tr ' ' '\n' | grep -Eq "^$1$"
```
Comment (Kijewski) [inline scripts/deps.sh]:
```sh
used() {
    [ "${DEPS#* $1 }" != "$DEPS" ];
}
```
DEPS has to be surrounded by spaces before the call.

**Comment by kaspar030** [comment]
@Kijewski pls don't comment at commit level, github eats those when the commit gets outdated.
I've changed ```used()``` to use shell substitution, that speeds up things quite a bit.

**Comment by kaspar030** [comment]
@Kijewski Hm, unfortunately that breaks wildcard matches.

File: scripts/deps.sh [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+shift
+
+DEPS="$*"
+
+use() {
+    DEPS="$DEPS $*"
+}
+
+used() {
+    DEPS=" $DEPS"
```
Comment (Kijewski) [inline scripts/deps.sh]: Trailing space is missing.

**Comment by Kijewski** [comment]
You can do a prefix match with `[ "${DEPS#* $1}" != "$DEPS" ]` (no space after `$1`)

File: scripts/deps.sh (hunk 6 above)
Comment (kaspar030) [inline scripts/deps.sh]: But it matches correctly like this.

Comment (Kijewski) [inline scripts/deps.sh]: Not for the last item in the list, or does it?

**Comment by haukepetersen** [comment]
About the windows users: I am not really thinking about people building RIOT on the command line (as most of us do it under linux/osx), but I am seriously thinking about IDE integration (-> which should be a major concern). As long as make is all you need, I imagine it easy to use RIOT in one of those large IDEs (Keil, IAR, etc.). But once we start to do more and more shell/python/whatever magic, it becomes harder and harder to integrate RIOT... 

I mean let's face it: for us the command line + Linux is what we prefer (and like) - but for 95% of the targeted, feature users of RIOT, a typical IDE under Windows is what counts (especially for industry users). So we should make sure, that we don't do anything with the make system that makes it less portable...

**Comment by OlegHahm** [comment]
Aren't we already using shell features in the build system at several points? And is there a GNU make for Windows apart from the cygwin version?

**Comment by authmillenon** [comment]
@OlegHahm yes, mingw.

**Comment by OlegHahm** [comment]
But mingw provides a POSIX shell as well, right?

**Comment by kaspar030** [comment]
I think I have a different approach (using make). Will close this for now.

**Comment by Kijewski** [comment]
Hm, I reopen this until I see your new approach. :)

**Comment by kaspar030** [comment]
Good idea, actually. gnrc's complicated conditional dependencies make simple solutions hard... ;)

**Comment by Kijewski** [comment]
Please see https://github.com/kaspar030/RIOT/pull/15

**Comment by kaspar030** [comment]
Nice! A lot faster now.

I'll rebase as soon as we're completely happy, as rebasing is a bitch when the underlying file completely disappears...

**Comment by OlegHahm** [comment]
Anyone still unhappy?

**Comment by Kijewski** [comment]
I'm happy, but probably the rules are outdated by now.

**Comment by kYc0o** [comment]
Needs a rebase, but cannot make it for this release...

**Comment by miri64** [comment]
@Kijewski do you have some sort of transformation script?

**Comment by miri64** [comment]
With #5891 merged, is this still required?

**Comment by miri64** [comment]
Postponed due to feature freeze (and please, can somebody answer my question?)

**Comment by Kijewski** [comment]
Superseded by #5891.

@kaspar030, please reopen if you disagree with our own PR. ;)

**Pull request closed without merging**